Honor whitelist settings when javac called via jar - fixes #2089

cprice404 opened the pull request:
This commit refactors a few things from the `jar` and
`javac` tasks in order to ensure that the "whitelisted"
settings from the user profile (`:local-repo`, `:mirrors`,
etc.) are honored when the `jar` task launches `javac`.

It also adds a test to validate the new behavior.

cprice404 commented:
This fixes #2089 for me.

hyPiRion commented:
Thanks for the PR =)

Seems like some existing tests with a cold cache broke. I'll have a look at it when I get the time, but I'd obviously be happy to get help with that if someone has time to look at it.

cprice404 commented:
hmmm, crud.  the tests passed for me locally.  Will try to repro the failure whenever I get a chance.

cprice404 commented:
It actually looks like it's my new test that is failing, and I am able to repro it locally now (on a different box from where I did the original work.  computers are awesome).  Will see if I can put up a fix.

cprice404 commented:
It looks like my test was using a maven2 mirror that no longer actually maintains any content.  I switched the test to use another mirror that actually has the clojure jar on it, and squashed that into my previous commit.  Tests pass for me locally now.

timvisher commented on test/leiningen/test/jar.clj:
```diff
@@ -134,3 +135,23 @@
+(deftest javac-launched-with-whitelisted-settings
+  (let [user-profile {:local-repo "foo/bar"
+                      :mirrors {"central" {:name "central"
+                                           :url "http://uk.maven.org/maven2"}}}
```
Is it just me or does relying on a maven repository on the internet seem like a bad idea in a unit test? Unless I'm misunderstanding the comments on the pr in general and this is just a value that could be arbitrary.

cprice404 replied:
I share the concern, but there are many tests in this repo that already rely implicitly on clojars and the "main" maven central repo.  You are correct that this is adding a dependency on one more internet site.  I'm open to other ideas on how to avoid that additional constraint in the test, but since a core part of leiningen is to resolve dependencies from the internets, and the bug that I opened was directly related to its ability to do that, the pros of having test coverage seemed to outweigh the cons of relying on one more mirror.

But, again, totally happy to revisit that if anyone has any better ideas!

timvisher replied:
Forgive me for butting in on something that I don't quite understand. :)

What about a `with-redefs` form that overrides the internal function that would fetch the repo from the internet and simply hard codes the return value?

hyPiRion replied:
Yeah, one should consider the leiningen tests as both unit tests and integration tests. It's really hard to test mirrors without actually using mirrors.

timvisher replied:
kk. carry on. :)

hyPiRion commented:
Thanks! =) This looks good – only thing that I guess may cause breakage here is moving `whitelist-keys` out of the jar namespace. I'll add in a def there with reference to the original thing to ensure backwards compatibility.
